lib/stddiff.ml

```diff
@@ -11,65 +11,71 @@ type 'a atomic_modification = { reference : 'a; current : 'a }
     [reference] is the value before and [current] is the value after the change
     occured. Use this type when there is no better representation available. *)
 
-type ('item, 'diff) map = {
-  same_map : 'item String_map.t;
-  changed_map : ('item, 'diff) entry String_map.t;
-}
-
-type ('same, 'change) maybe_changed = Same of 'same | Changed of 'change
 type 'item atomic_entry = ('item, 'item atomic_modification) entry
+type ('same, 'change) maybe_changed = Same of 'same | Changed of 'change
 
-type 'same maybe_changed_atomic =
-  ('same, 'same atomic_modification) maybe_changed
+module List_ = struct
+  type ('a, 'diff) t = ('a, ('a, 'diff) entry) maybe_changed list
 
-type 'same maybe_changed_atomic_entry =
-  ('same, 'same atomic_entry) maybe_changed
+  let diff ~diff_one ~reference ~current =
+    let rec aux reference current (acc, all) =
+      match (reference, current) with
+      | [], [] -> (List.rev acc, all)
+      | hd :: tl, [] -> aux tl [] (Changed (Removed hd) :: acc, false)
+      | [], hd :: tl -> aux [] tl (Changed (Added hd) :: acc, false)
+      | hd :: tl, hd' :: tl' -> (
+          let res = diff_one hd hd' in
+          match res with
+          | Same same -> aux tl tl' (Same same :: acc, true && all)
+          | Changed change ->
+              aux tl tl' (Changed (Modified change) :: acc, false && all))
+    in
+    let list_diff, all_same = aux reference current ([], true) in
+    if all_same then Same reference else Changed list_diff
+end
 
-type ('same, 'diff) option_ = ('same option, ('same, 'diff) entry) maybe_changed
-type 'same atomic_option = ('same, 'same atomic_modification) option_
-type ('same, 'diff) list_ = ('same list, 'diff list) maybe_changed
+module Option_ = struct
+  type ('a, 'diff) t = ('a option, ('a, 'diff) entry) maybe_changed
```

Comment:
This one is wrapped in a `maybe_changed` but the `List` one isn't. We should probably stick to the same convention here.

I don't think `M.t` should wrap into a maybe changed. The `diff` function could return an `(option, t) maybe_changed` but the type itself is the diff type, i.e. the type that represents the change.